python/guide/modules/autograd.md

```diff
@@ -1,178 +1,260 @@
-the fly. It's as if MXNet turned on a recording device and captured the exact
-path by which each variable was generated.
+Only operations that we want recorded are in the scope of the `autograd.record` context (since there is a computational overhead), and `autograd` should now have constructed a graph of these operations ready for the backward pass. We start the backward pass by calling the `.backward` method on the quantity of interest, which in this case is `loss` since were trying to calculate the gradient of the loss with respect to the parameters.
 
-Note that building the computation
-graph requires a nontrivial amount of computation. So MXNet will *only* build
-the graph when explicitly told to do so. This happens by placing code inside a
-``with autograd.record():`` block.
+Remember: if `loss` isn't a single scalar value (e.g. could be a loss for each sample, rather than for whole batch) a `sum` operation will be applied implicitly before starting the backward propogation, and the gradients calculated will be of this `sum` with respect to the parameters.
 
 ```{.python .input  n=4}
-with autograd.record():
-    y = 2 * nd.dot(x.T, x)
-print(y)
+loss.backward()
 ```
 
-Since the shape of `x` is (4, 1), `y` is a scalar. Next, we can automatically
-find the gradient by calling the `backward` function. It should be noted that if
-`y` is not a scalar, MXNet will first sum the elements in `y` to get the new
-variable by default, and then find the gradient of the variable with respect to
-`x`.
+And that's it! All the `autograd` magic is complete. We should now have gradients for each parameter of the network, which will be used by the optimizer to update the parameter values for improved performance. Check out the gradients of the first layer for example:
 
 ```{.python .input  n=5}
-y.backward()
+net[0].weight.grad()
 ```
 
-The gradient of the function $y = 2\mathbf{x}^{\top}\mathbf{x}$ with respect to
-$\mathbf{x}$ should be $4\mathbf{x}$. Now let's verify that the gradient
-produced is correct.
+## Advanced: Switching between training vs inference modes
+
+Some neural network layers behave differently depending on whether you're training the network or running it for inference. One example is `Dropout`, where activations are set to 0 at random during training, but remain unchanged during inference. Another is `BatchNorm`, where local batch statistics are used to normalize while training, but global statistics are used during inference.
+
+With MXNet Gluon, `autograd` is critical for switching between training and inference modes. As the default, networks will run in inference mode. While `autograd` is recording though, networks will run in training mode. Operations under the `autograd.record()` context scope are an example of this.
+
+Creating a network of a single `Dropout` block will demonstrate this.
 
 ```{.python .input  n=6}
-print((x.grad - 4 * x).norm().asscalar() == 0)
-print(x.grad)
-```
+import mxnet as mx
```

Comment:
2 spaces after

Reply:
same PEP 8 comments on the rest of the code

Reply:
done.